src/Blammo/Logging/Internal/LoggerLogAction.hs

```diff
@@ -0,0 +1,19 @@
+module Blammo.Logging.Internal.LoggerLogAction (loggerLogAction) where
+
+import Prelude
+
+import Blammo.Logging.Internal.LogAction
+import Blammo.Logging.Internal.Logger (Logger (..))
+import Blammo.Logging.Logger
+import Control.Monad (when)
+import Control.Monad.Logger.Aeson hiding (LoggingT (..), filterLogger)
+
+loggerLogAction :: Logger -> LogAction IO
+loggerLogAction logger =
+  LogAction $ \loc source level msg ->
+    when (lShouldLog logger source level) $
+      defaultOutputWith options loc source level msg
+ where
+  options = defaultOutputOptions $
+    \logLevel bytes ->
+      pushLogStrLn logger $ toLogStr $ getLoggerReformat logger logLevel bytes
```

Comment:
I don't think `LogAction m` is actually very useful. Why not do something simpler:

```hs
runLogAction
  :: ToLogStr msg => Logger -> Loc -> LogSource -> LogLevel -> msg -> IO ()
runLogAction logger lloc source level msg ->
    when (lShouldLog logger source level) $
      defaultOutputWith options loc source level msg
 where
  options = defaultOutputOptions $
    \logLevel bytes ->
      pushLogStrLn logger $ toLogStr $ getLoggerReformat logger logLevel bytes
```

Then in my other comment, it's similarly simplified:

```hs
runLoggerLoggingT
  :: (MonadUnliftIO m, HasLogger env) => env -> LoggingT m a -> m a
runLoggerLoggingT env (LoggingT f) =
  f (runLogAction logger) `finally` flushLogStr logger
 where
  logger = env ^. loggerL
```
I think you'd drop about 220 lines out of this PR, lose all your annoying CPP and older-ghc fuss -- with no difference in functionality.

We would still want to provide something for users of the library to derive `MonadLogger`, but I think that can also be simpler:

```hs
newtype ReaderLogger m a = ReaderLogger (m a)
  deriving newtype
    ( Functor
    , -- ...
    )

instance (MonadReader env m, HasLogger env) => MonadAWS (ReaderLogger m) where
  monadLoggerLog loc source level msg = do
    logger <- view loggerL
    runLogAction logger loc source level msg
```

(This is just the same pattern I use most places, e.g. [`amazonka-mtl`](https://hackage.haskell.org/package/amazonka-mtl-0.1.1.0/docs/src/Control.Monad.AWS.ViaReader.html#line-108))